src/conversion/viewconversiondispatcher.js

```diff
@@ -137,99 +159,179 @@ export default class ViewConversionDispatcher {
+		if ( viewItem.is( 'element' ) ) {
+			this.fire( 'element:' + viewItem.name, data, this.conversionApi );
+		} else if ( viewItem.is( 'text' ) ) {
+			this.fire( 'text', data, this.conversionApi );
 		} else {
-			this.fire( 'documentFragment', data, consumable, this.conversionApi );
+			this.fire( 'documentFragment', data, this.conversionApi );
 		}
 
-		// Handle incorrect `data.output`.
-		if ( data.output && !( data.output instanceof ModelNode || data.output instanceof ModelDocumentFragment ) ) {
+		// Handle incorrect conversion result.
+		if ( data.modelRange && !( data.modelRange instanceof ModelRange ) ) {
 			/**
 			 * Incorrect conversion result was dropped.
 			 *
-			 * Item may be converted to either {@link module:engine/model/node~Node model node} or
-			 * {@link module:engine/model/documentfragment~DocumentFragment model document fragment}.
+			 * {@link module:engine/model/range~Range Model range} should be a conversion result.
 			 *
 			 * @error view-conversion-dispatcher-incorrect-result
 			 */
-			log.warn( 'view-conversion-dispatcher-incorrect-result: Incorrect conversion result was dropped.', [ input, data.output ] );
-
-			return null;
+			throw new CKEditorError( 'view-conversion-dispatcher-incorrect-result: Incorrect conversion result was dropped.' );
 		}
 
-		return data.output;
+		return { modelRange: data.modelRange, cursorPosition: data.cursorPosition };
 	}
 
 	/**
 	 * @private
 	 * @see module:engine/conversion/viewconversiondispatcher~ViewConversionApi#convertChildren
 	 */
-	_convertChildren( input, consumable, additionalData ) {
-		const writer = this.conversionApi.writer;
-		const documentFragment = writer.createDocumentFragment();
+	_convertChildren( viewItem, cursorPosition ) {
+		const modelRange = new ModelRange( cursorPosition );
+		let nextCursorPosition = cursorPosition;
+
+		for ( const viewChild of Array.from( viewItem.getChildren() ) ) {
+			const result = this._convertItem( viewChild, nextCursorPosition );
+
+			if ( result.modelRange instanceof ModelRange ) {
+				modelRange.end = result.modelRange.end;
+				nextCursorPosition = result.cursorPosition;
+			}
+		}
+
+		return { modelRange, cursorPosition: nextCursorPosition };
+	}
+
+	/**
+	 * @private
+	 * @see module:engine/conversion/viewconversiondispatcher~ViewConversionApi#splitToAllowedParent
+	 */
+	_splitToAllowedParent( element, cursorPosition ) {
+		// Try to find allowed parent.
+		const allowedParent = this.conversionApi.schema.findAllowedParent( element, cursorPosition );
+
+		// When there is no parent that allows to insert element then return `null`.
+		if ( !allowedParent ) {
+			return null;
+		}
+
+		// When current position parent allows to insert element then return this position.
+		if ( allowedParent === cursorPosition.parent ) {
+			return { position: cursorPosition };
+		}
+
+		// When allowed parent is in context tree.
+		if ( this._modelCursor.parent.getAncestors().includes( allowedParent ) ) {
+			return null;
```

Comment:
Are you sure that null should be return here? Shouldn't we break it as much as possible in such case, and let `insertContent` do the rest?

Reply:
When insertContent is inserting conversion result it means that it is a pasting. Pasting is converted in `$clipboardHolder` context.

We do not allow anywhere to create invalid (from schema perspective) model tree. The idea of splitting was introduced to avoid creating invalid tree.